### ISSUE TITLE
Fully Conv Nets - Architecture Design and Training

### ISSUE THREAD

**Issue by rlee3359** [QUVA-Lab/escnn#4]
Hi there! Thanks for the library, really enjoying using it so far, and it seems really promising for my application!

I had a couple of questions regarding design and training choices for fully conv nets, would greatly appreciate some help! For context, I'm trying to track points in an image by training the network output heatmap against masks with 2D gaussians located at the label co-ordinates, with BCE + Sigmoid.

1. Training. When training such a model, how should the output of the network and the label be handled? 
I'm using `output_type = nn.FieldType(gc, 1*[gc.trivial_repr` for the final layer of the network. Do I need to cast the label to the same GeometricTensor:
`y = model.output_type(y)           # (results in an error in BCELoss - no attribute 'numel')`
Or do I pull out the torch tensor of the output?
`y_hat = y_hat.tensor                  # (this trains but seems to break the equivariance of the model)`

To test, I'm just training the model on one image and then testing the equivariance by rotating that one example and visualizing the output (using an architecture and viz setup similar to the animation example in the readme).

2. Upsampling: I noticed that R2Upsampling doesn't support a 'size' parameter as well as 'scale_factor' like torch.nn.Upsample does, is there a reason this is not implemented? Codewise it seems like a straightforward addition to the wrapper around torchs interpolate, but I thought maybe it might affect equivariance if the scaling factor was not perfect integer multiples, is that correct? I'm trying to make my output a heatmap with a specific shape (for example the same as the input shape or an specific down-scaling of the input shape). If this is not possible, I'll have to design the layers very deliberately to achieve a specific output size, I think.

3. On that note, I was wondering about conv layer design choices. I read through the paper and the documentation but I'm still not sure about how to choose kernel_size, stride and padding in the best way to preserve equivariance.
I did notice that [the mnist example](https://github.com/QUVA-Lab/escnn/blob/master/examples/model.ipynb) mentioned that stride 2 should not be used with odd kernel sizes (unless image padding is added), could you expand on this? What combination of kernel_size and stride is safe to use? I'm trying to re-create a model that has kernel of 5 and stride 2, but is it best to modify my network to avoid odd kernel sizes altogether? The example also mentioned that adding padding can allow the network to observe rotation artefacts and thus breaks equivariance somewhat. Is adding no padding safe? Sorry if that's a redundant question, I just thought I'd clarify to be sure :)

4. Lastly, I was curious about image padding and the MaskModule. Is it best practice to use one of these approaches? For now I'm manually masking everything outside the "image circle", is this strictly necessary? Is the mask module or padding out the image to prevent image corner loss better? 

Sorry for the long post! Really appreciate any help I can get with these questions. Thanks so much!

**Comment by rlee3359** [comment]
For reference, this is before training:

![before_training](https://user-images.githubusercontent.com/15622840/178977700-6dcf503a-f949-41ac-8208-d095bce1a270.gif)

After training:
![after_training](https://user-images.githubusercontent.com/15622840/178977736-b1dcff85-192d-4ffc-a6e4-a8fda90ec7b6.gif)

Where the network is just localising the lower left corner of the shape. Perhaps this is working, and the filters are just being activated by the other corners (that happen to look like similar but rotated versions). I'm using rot2dOnR2(N=8) for this model, and as the rotation passes multiples of 45, it does seem to have the same output.  Does this look like expected behaviour?

**Comment by Gabri95** [comment]
Hi @rlee3359 

I am happy to read you found our work useful!

Let me try to answer your questions:

1. No, you should use `y_hat.tensor` to extract the torch.Tensor and then use it as usual to compute the loss
   The equivariance break is probably caused by the training procedure, see next points

> To test, I'm just training the model on one image and then testing the equivariance by rotating that one example and visualizing the output (using an architecture and viz setup similar to the animation example in the readme).

If you are trying to achieve equivariace to N>4 rotations (i.e. to angles smaller than 90 degrees), this setup is likely to break equivariance. Unfortunately, since images are sampled on a squared grid, only equivariance to 90 degrees rotations can be perfect. That means any other model will only be *approximately* equivariant. However, if the training set is not "symmetric", i.e. does not actually contain rotated images, the optimization procedure can leverage this approximation to "overfit" on the fewer orientations present in the data, thereby increasing significantly the equivariance error.
Let me emphasise that this is not a problem if you only want N=4 rotations equivariance.

The presence of padding is another source of error. If the model is trained on augmented/rotated data, it tends to ignore these sources of noise since they are not useful for the task. Conversely, if the dataset is aligned in a single orientation, these sources of noise provide the model with enough information to break the equivariance and "specialize" on such orientation.

2. No there is no specific reason for not supporting those features, I just did not need them so far. I will include them in a future update, thanks for pointing this out!

3.  Check Fig 2 [here](https://arxiv.org/abs/2004.09691).  Generally speaking, it's better to use odd-sized filters and odd sized input images. In this way you can use strided convolution safely.
      Regarding the padding, see comment above. In other words, padding is an additional source of equivariance error; having no padding can help but it is not sufficient. Having an augmented dataset is generally the most important part. 

5.  The use of masked inputs is important to create a visualization like yours and/or to manually augment the training set with rotations. If your data naturally comes with rotated patterns and you only care about the final performance, you probably don't need to use this layer.

I hope this helps!
Gabriele

**Comment by rlee3359** [comment]
Thanks for the detailed response, that was super helpful.

Appreciate the clarification about training. Definitely I will have more rotations/augmentations when I apply these, I now see why my simple test setup would not work. I've changed my architecture based on your feedback and it's definitely improved. The odd-size image made a huge difference!

I implemented the size arg for Upsampling since I needed it anyways, with a pull-request here #5 in case that is helpful to you at all.

Thanks again!
